aten/src/ATen/native/BatchLinearAlgebra.cpp

```diff
@@ -206,6 +328,67 @@ extern "C" void strsm_(char *side, char *uplo, char *trans, char *diag, int *n,
 namespace at {
 namespace meta {
 
+TORCH_META_FUNC(linalg_ldl_factor_ex)
+(const Tensor& self, bool upper, bool hermitian, bool check_errors) {
+  at::native::squareCheckInputs(self, "torch.linalg.ldl_factor_ex");
+  at::native::checkFloatingOrComplex(self, "torch.linalg.ldl_factor_ex");
+
+  auto ndim = self.dim();
+  auto n = self.size(-2);
+
+  // prefer column major strides
+  auto ldu_strides = at::detail::defaultStrides(self.sizes());
+  ldu_strides[ndim - 2] = 1;
+  ldu_strides[ndim - 1] = std::max<int64_t>(1, n);
+  set_output(0, self.sizes(), ldu_strides, self.options(), {}); // factors
+
+  auto pivots_shape =
+      IntArrayRef(self.sizes().data(), ndim - 1); // self.shape[:-1]
+  set_output(
+      1, pivots_shape, {}, self.options().dtype(ScalarType::Int), {}); // pivots
+
+  auto info_shape =
+      IntArrayRef(self.sizes().data(), ndim - 2); // self.shape[:-2]
+  set_output(
+      2, info_shape, {}, self.options().dtype(ScalarType::Int), {}); // info
+}
+
+TORCH_META_FUNC(linalg_ldl_solve)
+(const Tensor& factors,
+ const Tensor& pivots,
+ const Tensor& B,
+ bool upper,
+ bool hermitian) {
```

Comment:
This flag is cheap in BLAS (hence solve_triangular). But ldl_solve LAPACK doesn't even have transpose flags and doesn't have left/right flags.

Reply:
What I outlined up there is a way to implement it at the `TORCH_IMPL_FUNC` level via a tad of extra preprocessing and postprocessing.

Reply:
I'd rather submit an issue after this PR because of "extra preprocessing and postprocessing". It could be done in the future hypothetical generic `factorize + solve` functions that wrap these factorization-specific solvers.

It's possible to pass a triangular matrix to `ldl_factor` and everything would still work as expected and we can't restrict the input to be strictly symmetric/Hermitian. Since passing triangular matrix is allowed then `A != A.mT` and it's not possible to reuse the same factorization for the "left" solver.

Reply:
So, in my opinion, this function should expect the inputs to be always symmetric / Hermitian according to the flag, and thus, its gradients should also be symmetric or Hermitian.

If that's the case, we should be allowed to do the thing proposed above although, of course, it needn't be in this PR. Could you leave a comment at the top of the function sayin that it'd be cool to implement that?

Now, about supporting triangular matrices, if you're keen on doing so, fine (although I'd prefer not to do so) but then you'd need to:
1. Document this properly, as it's not the expected behaviour and it's currently undocumented.
2. Be careful with the implementation of the gradients. This function will oddly return upper/lower triangular gradients, as opposed to what `linalg.cholesky` and `linalg.eigh` do. We should probably also document this, as it's a bit odd.

Reply:
Both `linalg.cholesky` and `linalg.eigh` work with triangular input:
```
In [1]: import torch

In [2]: a = torch.randn(3, 3)

In [3]: a = a @ a.mT

In [4]: torch.allclose(torch.linalg.cholesky(a), torch.linalg.cholesky(a.tril()))
Out[4]: True

In [5]: [torch.allclose(x, y) for x, y in zip(torch.linalg.eigh(a), torch.linalg.eigh(a.tril()))]
Out[5]: [True, True]
```

Reply:
If you pass a triangular output the gradient and forward AD will be incorrect

Reply:
Note that this was one of the points raised that led to the removal of the `upper` argument in `eigh` and `eigvalsh` in the array API https://github.com/data-apis/array-api/issues/217

Reply:
For Cholesky and eigh it is symmetrized and correct. Previously PyTorch (pre ~1.5) would return triangular derivatives for Cholesky, but it was changed.

```python
In [1]: import torch

In [2]: a = torch.randn(3, 3)

In [3]: a = a @ a.mT

In [4]: a_tril = a.tril()

In [5]: g = torch.randn(3, 3)

In [6]: a_tril.requires_grad_(True)
Out[6]:
tensor([[ 9.1411,  0.0000,  0.0000],
        [-6.0545,  4.1667,  0.0000],
        [-3.7081,  2.3873,  3.4195]], requires_grad=True)

In [7]: L = torch.linalg.cholesky(a_tril)

In [8]: L
Out[8]:
tensor([[ 3.0234,  0.0000,  0.0000],
        [-2.0025,  0.3957,  0.0000],
        [-1.2264, -0.1736,  1.3730]], grad_fn=<LinalgCholeskyExBackward0>)

In [9]: L.backward(g)

In [10]: a_tril.grad
Out[10]:
tensor([[ 0.0103,  0.9693, -0.4156],
        [ 0.9693,  1.7575, -0.7359],
        [-0.4156, -0.7359,  0.6651]])
```